File: app/src/main/java/com/vikho305/isaho220/outstanding/data/api/PostApi.java
package com.vikho305.isaho220.outstanding.data.api;

import android.content.Context;

import com.rx2androidnetworking.Rx2AndroidNetworking;
import com.vikho305.isaho220.outstanding.data.AuthInfo;
import com.vikho305.isaho220.outstanding.data.Comment;
import com.vikho305.isaho220.outstanding.data.Post;
import com.vikho305.isaho220.outstanding.data.User;
import com.vikho305.isaho220.outstanding.data.repositories.PreferenceRepository;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.List;

import io.reactivex.Single;

public class PostApi {
    private static final String ROOT_URL = "https://outstanding-server.herokuapp.com/post";

    private PreferenceRepository preferences;

    public PostApi(Context context) {
        preferences = new PreferenceRepository(context);
    }

    public Single<Post> getPost(String postId) {
        return Rx2AndroidNetworking.get(ROOT_URL + "/" + postId)
                .addHeaders("Authorization", "Bearer " + preferences.getAuthToken())
                .build()
                .getObjectSingle(Post.class);
    }

    public Single<List<Post>> getPostsInRadius(double radius) {
        return Rx2AndroidNetworking.get(ROOT_URL + "/all/" + radius)
                .addHeaders("Authorization", "Bearer " + preferences.getAuthToken())
                .build()
                .getObjectListSingle(Post.class);
    }

    public Single<List<Comment>> getComments(String postId, int page) {
        return Rx2AndroidNetworking.get(ROOT_URL + "/" + postId + "/comments/" + page)
                .addHeaders("Authorization", "Bearer " + preferences.getAuthToken())
                .build()
                .getObjectListSingle(Comment.class);
    }

    public Single<String> postPost(String title, String text, String media, String mediaType,
                                   double latitude, double longitude) {
        JSONObject jsonParameters = new JSONObject();
        try {
            jsonParameters.put("title", title);
            jsonParameters.put("text", text);
            jsonParameters.put("media", media);
            jsonParameters.put("media_type", mediaType);
            jsonParameters.put("latitude", latitude);
            jsonParameters.put("longitude", longitude);
        } catch (JSONException e) {
            e.printStackTrace();
        }

        return Rx2AndroidNetworking.post(ROOT_URL)
                .addHeaders("Authorization", "Bearer " + preferences.getAuthToken())
                .addJSONObjectBody(jsonParameters)
                .build()
                .getStringSingle();
    }

    public Single<String> postComment(String postId, String text) {
        JSONObject jsonParameters = new JSONObject();
        try {
            jsonParameters.put("text", text);
        } catch (JSONException e) {
            e.printStackTrace();
        }

        return Rx2AndroidNetworking.post(ROOT_URL + "/" + postId + "/comment")
                .addHeaders("Authorization", "Bearer " + preferences.getAuthToken())
                .addJSONObjectBody(jsonParameters)
                .build()
                .getStringSingle();
    }

    public Single<String> likePost(String postId) {
        return Rx2AndroidNetworking.post(ROOT_URL + "/" + postId + "/like")
                .addHeaders("Authorization", "Bearer " + preferences.getAuthToken())
                .build()
                .getStringSingle();
    }

    public Single<String> dislikePost(String postId) {
        return Rx2AndroidNetworking.post(ROOT_URL + "/" + postId + "/dislike")
                .addHeaders("Authorization", "Bearer " + preferences.getAuthToken())
                .build()
                .getStringSingle();
    }

    public Single<String> unlikePost(String postId) {
        return Rx2AndroidNetworking.delete(ROOT_URL + "/" + postId + "/like")
                .addHeaders("Authorization", "Bearer " + preferences.getAuthToken())
                .build()
                .getStringSingle();
    }

    public Single<String> undislikePost(String postId) {
        return Rx2AndroidNetworking.delete(ROOT_URL + "/" + postId + "/dislike")
                .addHeaders("Authorization", "Bearer " + preferences.getAuthToken())
                .build()
                .getStringSingle();
    }
}

File: app/src/main/java/com/vikho305/isaho220/outstanding/data/api/PostApi.kt
package com.vikho305.isaho220.outstanding.data.api

import android.content.Context
import com.rx2androidnetworking.Rx2AndroidNetworking
import com.vikho305.isaho220.outstanding.data.Comment
import com.vikho305.isaho220.outstanding.data.Post
import com.vikho305.isaho220.outstanding.data.repositories.PreferenceRepository
import io.reactivex.Single
import org.json.JSONException
import org.json.JSONObject

class PostApi(context: Context) : Api(context,"post") {
    fun getPost(postId: String): Single<Post> {
        return get(postId, true).getObjectSingle(Post::class.java)
    }

    fun getPostsInRadius(radius: Double): Single<List<Post>> {
        return get("all/$radius", true).getObjectListSingle(Post::class.java)
    }

    fun getComments(postId: String, page: Int): Single<List<Comment>> {
        return get("$postId/comments/$page", true).getObjectListSingle(Comment::class.java)
    }

    fun postPost(title: String?, text: String?, media: String?, mediaType: String?,
                 latitude: Double, longitude: Double): Single<String> {
        val jsonParameters = JSONObject()

        try {
            jsonParameters.put("title", title)
            jsonParameters.put("text", text)
            jsonParameters.put("media", media)
            jsonParameters.put("media_type", mediaType)
            jsonParameters.put("latitude", latitude)
            jsonParameters.put("longitude", longitude)
        } catch (e: JSONException) {
            e.printStackTrace()
        }

        return post("", true, jsonParameters).stringSingle
    }

    fun postComment(postId: String, text: String?): Single<String> {
        val jsonParameters = JSONObject()

        try {
            jsonParameters.put("text", text)
        } catch (e: JSONException) {
            e.printStackTrace()
        }

        return post("$postId/comment", true, jsonParameters).stringSingle
    }

    fun likePost(postId: String): Single<String> {
        return post("$postId/like", true).stringSingle
    }

    fun dislikePost(postId: String): Single<String> {
        return post("$postId/dislike", true).stringSingle
    }

    fun unlikePost(postId: String): Single<String> {
        return delete("$postId/like", true).stringSingle
    }

    fun undislikePost(postId: String): Single<String> {
        return delete("$postId/dislike", true).stringSingle
    }
}